### PR TITLE
Blockprod unit testing

### PR DESCRIPTION
**Pull request by alfiedotwtf**
Adding Proof of Stake testing to `blockprod`.

### CODE REVIEW

**Comment by alfiedotwtf** [comment]
As a last step, or to be done sometime after this PR as it's low priority - https://github.com/mintlayer/mintlayer-core/issues/957

File: blockprod/src/detail/tests.rs [hunk 1]
```diff
@@ -185,26 +426,128 @@ async fn stop_job_non_existent_job(#[case] seed: Seed) {
...
+    mock_job_manager
+        .expect_stop_all_jobs()
+        .times(1)
+        .returning(move || Ok(return_value));
+
+    block_production.set_job_manager(mock_job_manager);
+
+    let result = block_production.stop_all_jobs().await;
+
+    assert!(matches!(result, Ok(_expected_value)));
```
Comment (azarovh) [inline blockprod/src/detail/tests.rs:506]: matches! is redundant

File: blockprod/src/detail/tests.rs [hunk 2]
```diff
@@ -308,74 +649,47 @@ async fn produce_block_multiple_jobs(#[case] seed: Seed) {
...
+    block_production.set_job_manager(mock_job_manager);
 
-    let jobs_count = block_production.job_manager.get_job_count().await.unwrap();
-    assert_eq!(jobs_count, 0, "Jobs count is incorrect");
+    let mut rng = make_seedable_rng(seed);
+    let job_key = JobKey::new(None, Id::new(H256::random_using(&mut rng)) as Id<GenBlock>);
+
+    let result = block_production.stop_job(job_key).await;
+
+    assert!(matches!(result, Ok(true)));
```
Comment (azarovh) [inline blockprod/src/detail/tests.rs:694]: assert_eq

**Review by azarovh — APPROVED** [review]

File: blockprod/src/detail/mod.rs [hunk 3]
```diff
@@ -474,4 +480,4 @@ fn generate_finalize_block_data(
 }
 
 #[cfg(test)]
-mod tests;
+pub mod tests;
```
Comment (TheQuantumPhysicist) [inline blockprod/src/detail/mod.rs:483]: This should never be done. Exposing test modules guarded by cfg(test) can lead to all kinds of problems, including the false perception that something is public, while it's not.

File: blockprod/src/detail/mod.rs [hunk 4]
```diff
@@ -94,15 +94,21 @@ impl BlockProduction {
         &self.time_getter
     }
 
+    /// Private method used for unit testing
+    #[allow(dead_code)]
+    fn set_job_manager(&mut self, job_manager_handle: JobManagerHandle) {
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply blockprod/src/detail/mod.rs:99]: Once you do that, you won't need to allow(dead_code)

File: blockprod/src/lib.rs [hunk 5]
```diff
@@ -94,18 +94,72 @@ pub fn make_blockproduction(
...
+        vrf::{VRFKeyKind, VRFPrivateKey},
+    };
     use mempool::{MempoolHandle, MempoolSubsystemInterface};
     use subsystem::Manager;
+    use test_utils::random::{make_seedable_rng, Seed};
 
     use super::*;
 
-    pub fn setup_blockprod_test() -> (Manager, Arc<ChainConfig>, ChainstateHandle, MempoolHandle) {
+    const MIN_STAKE_POOL_PLEDGE: Amount = Amount::from_atoms(40_000 * Mlt::ATOMS_PER_MLT);
```
Comment (TheQuantumPhysicist) [inline blockprod/src/lib.rs:127]: Please take this value from `ChainConfig::min_stake_pool_pledge`.

**Review by iljakuklic — APPROVED** [review]

File: blockprod/src/detail/tests.rs (hunk 1 above)
Comment (alfiedotwtf) [reply blockprod/src/detail/tests.rs:506]: If I remove `matches!` here, `assert_eq!` will fail to compile. I've changed it into a vanilla `match`.

Comment (iljakuklic) [reply blockprod/src/detail/tests.rs]: Would it be solved by adding the `Eq` impl to the error type, if it's not there already?

Comment (alfiedotwtf) [reply blockprod/src/detail/tests.rs]: I thought about doing this, but didn't want to pollute the code and binary just for a single test. It's doing what it needs to without the `matches!` now, but if `Eq` is warranted, then I'm happy to add it.

Comment (alfiedotwtf) [reply blockprod/src/detail/tests.rs]: Cool, no problem.

**Pull request merged**